ownCloud/Client/Sharing/PublicLinkTableViewController.swift

```diff
@@ -0,0 +1,273 @@
+				if canEdit(share: share) {
+					shareRows.append( StaticTableViewRow(rowWithAction: { [weak self] (_, _) in
+						guard let self = self else { return }
+						let editPublicLinkViewController = PublicLinkEditTableViewController(style: .grouped)
+						editPublicLinkViewController.share = share
+						editPublicLinkViewController.core = self.core
+						editPublicLinkViewController.item = self.item
+						self.navigationController?.pushViewController(editPublicLinkViewController, animated: true)
+					}, title: share.name!, subtitle: share.permissionDescription(), accessoryType: .disclosureIndicator) )
+				} else {
+					shareRows.append( StaticTableViewRow(rowWithAction: nil, title: share.name!, subtitle: share.permissionDescription(), accessoryType: .none) )
+				}
+			}
+		} else {
+			shareRows.append( StaticTableViewRow(buttonWithAction: { [weak self] (_, _) in
+				self?.addPublicLink()
+			}, title: "Create Public Link".localized, style: StaticTableViewRowButtonStyle.plain))
+		}
+
+		let sectionType = "share-section-\(String(type.rawValue))"
+		if let section = self.sectionForIdentifier(sectionType) {
+			self.removeSection(section)
+		}
+		let section : StaticTableViewSection = StaticTableViewSection(headerTitle: title, footerTitle: nil, identifier: sectionType, rows: shareRows)
+		self.addSection(section, animated: false)
+	}
+
+	func addShareSections() {
+		OnMainThread {
+			self.addSectionFor(type: .link, with: "Public Links".localized)
+		}
+	}
+
+	func removeShareSections() {
+		OnMainThread {
+			let types : [OCShareType] = [.link]
+			for type in types {
+				let identifier = "share-section-\(String(type.rawValue))"
+				if let section = self.sectionForIdentifier(identifier) {
+					self.removeSection(section)
+				}
+			}
+		}
+	}
+
+	func resetTable(showShares : Bool) {
+		removeShareSections()
+		if shares.count > 0 && showShares {
+			messageView?.message(show: false)
+		}
+		self.addShareSections()
+	}
+
+	func handleEmptyShares() {
+		if shares.count == 0 {
+			OnMainThread {
+				self.resetTable(showShares: false)
+			}
+		}
+	}
+
+	// MARK: - Private Link Section
+
+	func addPrivateLinkSection() {
+		let identifier = "private-link-section"
+		if let section = self.sectionForIdentifier(identifier) {
+			self.removeSection(section)
+		}
+
+		let footer = "Only collaborators can use this link. Use it as a permanent link to point to this resource".localized
+
+		OnMainThread {
+			let section = StaticTableViewSection(headerTitle: nil, footerTitle: footer, identifier: "private-link-section")
+			var rows : [StaticTableViewRow] = []
+
+			self.core?.retrievePrivateLink(for: self.item, completionHandler: { (error, url) in
+
+				guard let url = url else { return }
+				if error == nil {
+					OnMainThread {
+						let privateLinkRow = StaticTableViewRow(headerRowWithAction: { [weak self] (row, _) in
+							guard let self = self else { return }
+							self.retrievePrivateLink(for: self.item, in: row)
```

Comment:
At this point, the private link URL is already known. As far as I know that URL doesn't change, so it shouldn't be necessary to retrieve it again in `retrievePrivateLink(for:in:)`. Instead, the private link URL can be "saved" in this block and copied from clipboard when run.